File: e2e/typescript/amqp.js
const { AmqpPublisher, AmqpSubscriber } = require("../../sdk/amqp/typescript/dist/index.js");

async function main() {
  const mode = process.argv[2] || "sub";
  const url = process.env.AMQP_URL || "amqp://guest:guest@localhost:5672/";
  const topic = process.env.TOPIC || "relaybus.alpha";
  const exchange = process.env.AMQP_EXCHANGE || "relaybus.events";
  const exchangeType = process.env.AMQP_EXCHANGE_TYPE || "topic";
  const queue = process.env.AMQP_QUEUE || `${topic}.queue`;

  if (mode === "sub") {
    const subscriber = await AmqpSubscriber.connect({
      url,
      exchange,
      exchangeType,
      queue,
      onMessage: (msg) => {
        console.log(`received id=${msg.id} topic=${msg.topic} payload=${msg.payload.toString()}`);
      }
    });

    await withTimeout(subscriber.start(topic), 30000);
    await subscriber.close();
    return;
  }

  const publisher = await AmqpPublisher.connect({ url, exchange, exchangeType, queue });

  await publisher.publish(topic, {
    topic,
    payload: Buffer.from("hello from typescript"),
    meta: { lang: "ts" }
  });

  await publisher.close();
  console.log("published");
}

function withTimeout(promise, ms) {
  return Promise.race([
    promise,
    new Promise((_, reject) =>
      setTimeout(() => reject(new Error("timeout waiting for message")), ms)
    )
  ]);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: e2e/python/amqp.py
from __future__ import annotations

import os
import sys

from relaybus_amqp import (
    AmqpPublisher,
    AmqpPublisherConnectConfig,
    AmqpSubscriber,
    AmqpSubscriberConnectConfig,
)
from relaybus_core import OutgoingMessage


def main() -> None:
    mode = sys.argv[1] if len(sys.argv) > 1 else "sub"
    url = os.getenv("AMQP_URL", "amqp://guest:guest@localhost:5672/")
    topic = os.getenv("TOPIC", "relaybus.alpha")
    exchange = os.getenv("AMQP_EXCHANGE", "relaybus.events")
    exchange_type = os.getenv("AMQP_EXCHANGE_TYPE", "topic")

    if mode == "sub":
        subscriber = AmqpSubscriber.connect(
            AmqpSubscriberConnectConfig(
                url=url,
                exchange=exchange,
                exchange_type=exchange_type,
                on_message=lambda msg: print(
                    f"received id={msg.id} topic={msg.topic} payload={msg.payload.decode()}"
                ),
                queue=topic,
            )
        )
        subscriber.start(topic)
        subscriber.close()
        return

    publisher = AmqpPublisher.connect(
        AmqpPublisherConnectConfig(url=url, exchange=exchange, exchange_type=exchange_type)
    )
    publisher.publish(topic, OutgoingMessage(topic=topic, payload=b"hello from python", meta={"lang": "py"}))
    publisher.close()
    print("published")


if __name__ == "__main__":
    main()
